File: CSE222 - Data Structures and Algorithms/hw/hw02/src/PairsOfSum.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int pairsOfSumLoop(int arr[], int size, int sum) {
    int count = 0;
    for (int i = 0; i < size; ++i) {
        int target = sum - arr[i];
        for (int j = i + 1; j < size; ++j)
            if (arr[j] == target)
                ++count;
    }
    return count;
}

int pairsOfSumRecursive(int arr[], int size, int sum) {
    int count = 0;
    if (size > 0) {
        int target = sum - arr[size - 1];
        for (int i = size - 2; i >= 0; --i)
            if (arr[i] == target)
                ++count;
        count += pairsOfSumRecursive(arr, size - 1, sum);
    }
    return count;
}

int main(void) {
    const int ARR_SIZE = 1000000;
    const int sum = 10;
    int vals[] = {-9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    
    int * arr = (int *) calloc(ARR_SIZE, sizeof(int));

    for (int i = 0; i < ARR_SIZE; ++i)
        arr[i] = vals[i % 20];

    for (int n = 1000; n <= ARR_SIZE; n *= 10) {
        clock_t begin = clock();
        pairsOfSumLoop(arr, n, sum); 
        // pairsOfSumRecursive(arr, 15, n - 1);
        clock_t end = clock();

        double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
        printf("Running time for problem size %-10.d: %.5f\n", n, time_spent);
    }

    free(arr);
}

File: cse222-data-structures-and-algorithms/hw/hw02/src/PairsOfSum.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

class PairsOfSum {
    public static void main(String[] args) {
        ArrayList<String> list = new ArrayList<>();
        for (int i = 10; i <= 10000; i *= 10)
            list.add(test(i));
        
        // print the result
        for (var r : list)
            System.out.println(r);
    }

    public static String test(int size) {
        int[] arr = createArray(size);
        
        long startTime = System.nanoTime();
        pairsOfSumLoop(arr, 25);
        // pairsOfSumRecursive(arr, 25, 0);
        long stopTime = System.nanoTime();

        return String.format("Running time for problem size %7d: %15.2f", size, (stopTime - startTime) / 1000.0);
    }

    public static void pairsOfSumLoop(int[] arr, int sum) {
        int count = 0;
        for (int i = 0; i < arr.length; ++i) {
            int target = sum - arr[i];
            for (int j = i + 1; j < arr.length; ++j)
                if (arr[j] == target)
                    ++count;
                    // System.out.printf("Pair (%d, %d)\n", arr[i], target);
        }
    }

    public static void pairsOfSumRecursive(int[] arr, int sum, int index) {
        int count = 0;
        if (index < arr.length - 1) {
            int target = sum - arr[index];
            for (int i = index + 1; i < arr.length; ++i)
                if (arr[i] == target)
                    ++count;
                    // System.out.printf("Pair (%d, %d)\n", arr[index], target);
            pairsOfSumRecursive(arr, sum, index + 1);
        }
    }

    public static int[] createArray(int size) {
        int[] arr = new int[size];

        try (Scanner scanner = new Scanner(new File("random.txt"))) {
            for (int i = 0; i < size && scanner.hasNext(); ++i)
                arr[i++] = scanner.nextInt();
            scanner.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        return arr;
    }

}
